protocol/src/protocol.semanticHighlighting.proposed.md

```diff
@@ -0,0 +1,88 @@
+#### Semantic Highlighting
+
+While the syntax highlighting are done on the client-side and can handle keywords, strings, and other low-level tokens from the grammar, it cannot adequately support complex coloring. Semantic highlighting information is calculated on the language server and pushed to the client as a notification.  This notification carries information about the ranges that have to be colored. The desired coloring details are given as TextMate scopes for each affected range. For the semantic highlighting information the following additions are proposed:
+
+_Client Capabilities_:
+
+Capability that has to be set by the language client if that can accept and process the semantic highlighting information received from the server.
+
+```ts
+/**
+ * The text document client capabilities.
+ */
+workspace: {
+
+	/**
+	 * `true` if the client has the semantic highlighting support for the text document. Otherwise, `false`. It is `false` by default.
+	 */
+	semanticHighlighting?: boolean;
+}
+```
+
+##### SemanticHighlighting Notification
+
+The `textDocument/semanticHighlighting` notification is pushed from the server to the client to inform the client about additional semantic highlighting information that has to be applied on the text document.
+
+_Notification_:
+
+* method: 'workspace/semanticHighlighting'
+* params: `SemanticHighlightingParams` defined as follows:
+
+```ts
+/**
+ * Parameters for the semantic highlighting (server-side) push notification.
+ */
+export interface SemanticHighlightingParams {
+
+	/**
+ 	* The URI for which semantic highlighting information is reported to.
+ 	*/
+	uri: string;
+
+	/**
+	 * An array of semantic highlighting information.
+	 */
+	lines: SemanticHighlightingInformation[];
+
+}
+
+/**
+ * Represents a semantic highlighting information that has to be applied on a specific line of the text document.
+ */
+export interface SemanticHighlightingInformation {
+
+	/**
+     * The zero-based line position in the text document.
+     */
+	line: number;
+
+	/**
+	 * An array of highlighting tokens for the current line in the text document.
+	 */
+	tokens: SemanticHighlightingToken[];
+
+}
+
+/**
+ * Representation of a single semantic highlighting token that has to be applied on a line of the text document.
+ */
+export interface SemanticHighlightingToken {
+
+	/**
+     * The zero-based character offset on the line in a text document where this token applies. If negative, defaults to `0`.
+	 * If greater than the length of the current line, defaults to the end of the line.
+     */
+	character: number;
+
+	/**
+	 * The length of the token. If negative, defaults to `0`. If the sum of the `character` and the `length` exceeds the length of the line,
+	 * it defaults back to the end of the line.
+	 */
+	length: number;
+
+	/**
+	 * An array of semantic highlighting scopes for the current token.
+	 */
+	scopes: string[];
```

Comment:
Documentation should point out that this corresponds to TextMate scopes. https://manual.macromates.com/en/language_grammars